src/Database/LSMTree/Common.hs

```diff
@@ -77,9 +85,11 @@ instance IOLike IO
 -- while a shared session will place all files under one directory.
 --
 type Session :: (Type -> Type) -> Type
-data Session m = Session {
-    sessionRoot  :: !FsPath
-  , sessionHasFS :: !(SomeHasFS m)
+data Session m = forall h. Typeable h => Session {
+    sessionRoot       :: !FsPath
+  , sessionHasFS      :: !(HasFS m h)
+  , sessionHasBufFS   :: !(HasBufFS m h)
+  , sessionHasBlockIO :: !(HasBlockIO m h)
```

Comment:
Yes ok nice. So we don't need to use "Some" for the HasFS, and we can see each of them use the same handle type, but we still hide the `h` param from the `Session m` type that is exposed in the API.

And I'm guessing it's `Typeable` because we're going to have to do a dynamic type check to recover the type?